src/crimson/os/seastore/omap_manager/btree/omap_btree_node_impl.cc

```diff
@@ -732,36 +810,71 @@ OMapLeafNode::make_balanced(omap_context_t oc, OMapNodeRef _right)
   );
 }
 
-
-omap_load_extent_iertr::future<OMapNodeRef>
-omap_load_extent(omap_context_t oc, laddr_t laddr, depth_t depth)
+OMapInnerNode::get_child_node_ret
+OMapInnerNode::get_child_node(
+  omap_context_t oc,
+  internal_const_iterator_t child_pt)
 {
-  ceph_assert(depth > 0);
-  if (depth > 1) {
-    return oc.tm.read_extent<OMapInnerNode>(
-        oc.t, laddr, OMAP_INNER_BLOCK_SIZE
-    ).handle_error_interruptible(
-      omap_load_extent_iertr::pass_further{},
-      crimson::ct_error::assert_all{ "Invalid error in omap_load_extent" }
-    ).si_then([](auto maybe_indirect_extent) {
-      assert(!maybe_indirect_extent.is_indirect());
-      assert(!maybe_indirect_extent.is_clone);
-      return seastar::make_ready_future<OMapNodeRef>(
-          std::move(maybe_indirect_extent.extent));
+  assert(get_meta().depth > 1);
+  child_pos_t<OMapInnerNode> child_pos(nullptr, 0);
+  auto laddr = child_pt->get_val();
+  auto next = child_pt + 1;
+  if (get_meta().depth == 2) {
+    auto ret = this->get_child<OMapLeafNode>(
+      oc.t, oc.tm.get_etvr(), child_pt.get_offset(), child_pt.get_key());
+    if (ret.has_child()) {
+      return ret.get_child_fut(
+      ).si_then([](auto extent) {
+	return extent->template cast<OMapNode>();
+      });
+    } else {
+      child_pos = ret.get_child_pos();
+    }
+    return omap_load_extent<OMapLeafNode>(
+      oc,
+      laddr,
+      get_meta().depth - 1,
+      child_pt->get_key(),
+      next == iter_cend()
+	? this->get_end()
+	: next->get_key()
+    ).si_then([chp=std::move(child_pos)](auto extent) mutable {
+      chp.link_child(extent.get());
```

Comment:
This looks different from the lba tree's implementation.

In the lba tree, `link_child()` is called atomically with `get_child()`. This is because in seastore, reads are allowed to be invoked concurrently. And if they are not atomic, it will be possible to cause `omap_load_extent()` and `link_child()` to be called multiple times for the same extent, which is illegal and result in undefined behavior.

Reply:
Ah, yes. Will correct it.

Reply:
Um, after a second thought, it seems that omap extents are not to be accessed concurrently, so maybe we can ignore this case and add some assertions to make sure concurrent access never happens? What do you think?

Reply:
>  it seems that omap extents are not to be accessed concurrently

I don't see where prevents concurrent accesses, but seems to me OSD (tri-mutex) allows multiple readers for an object at the same time.

Reply:
> > it seems that omap extents are not to be accessed concurrently
> 
> I don't see where prevents concurrent accesses, but seems to me OSD (tri-mutex) allows multiple readers for an object at the same time.

Ah, yes, sorry, my mistake.

Reply:
> Um, it seems that we need to add a callback parameter to TransactionManager::read_extent(), TransactionManager::read_pin() and TransactionManager::pin_to_extent() to tackle this issue.

Ah it looks more complicated than I thought, `get_pin()` is unconditionally asynchronous.

Reply:
Seems to me the key difference is that if the child omap node is unlinked/absent, we cannot atomically construct the omap node before query the lba tree.